skimmer/skimmer_utils.py

```diff
@@ -472,4 +488,32 @@ def apply_variation(events, variation):
                 k,
             )
 
-        return events
\ No newline at end of file
+        return events
+    
+
+def apply_scale_variations(events):
+    '''
+    # Get up/down variations envelope, following definition
+    # here: https://github.com/TreeMaker/TreeMaker/blob/7a81115566ed1f2206eb4d447c9c7ba0870d88d0/Utils/src/PDFWeightProducer.cc#L167
+    
+    # Adds "ScaleWeight_up" and "ScaleWeight_down" branches, plus stores the integrals
+    # for nominal and up/down variations in the cutflow
+    '''
+
+    envelope_up = ak.max(events.ScaleWeights[:,[i for i in range(9) if i not in (5, 7)]], axis=-1)
+    envelope_down = ak.min(events.ScaleWeights[:, [i for i in range(9) if i not in (5, 7)]], axis=-1)
+
+    # Calculate central norm and variations
+    if is_tree_maker(events):
+        sum_w_nominal = ak.sum(events.Weight)
+        sum_w_up = ak.sum(events.Weight * envelope_up)
+        sum_w_down = ak.sum(events.Weight * envelope_down)
+    else:
+        sum_w_nominal = ak.sum(events.genWeight)
+        sum_w_up = ak.sum(events.genWeight * envelope_up)
+        sum_w_down = ak.sum(events.genWeight * envelope_down)
+
+    events["ScaleWeight_up"] = envelope_up
```

Comment:
Not sure I understand, storing the weights is exactly what I'm doing

Reply:
The weights are `events.Weight * envelope_up`, not just `envelope_up` right? Otherwise, the sum of weight to normalize correctly wouldn't be `ak.sum(events.genWeight * envelope_up)`.
What I mean is that instead of having to multiply two sets of weights, it would be easier to just have `Weight` for nominal, and `WeightUp` and `WeightDown` for the variations.

Reply:
No, because we have different independent variations of the weights i.e., PDF, PS, Scales...

Reply:
Clarified verbally :)